text/0000-unreservations-2018.md

```diff
@@ -0,0 +1,302 @@
+[drawbacks]: #drawbacks
+
+The only drawback is that we're not able to use each listed word as a keyword
+in the future, without a reservation in a new edition, if we realize that we
+made a mistake.
+
+See the rationale for potential risks with each keyword.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+There's only one alternative: Not unreserving all listed / some keywords.
+
+Not unreserving a keyword would make the word unavailable for use as an
+identifier.
+
+## General policy around unreservations
+
+This RFC establishes a general rationale and policy for keyword unreservation:
+*If we are not using a keyword for anything in the language, and we are sure
+that we have no intention of using the keyword in the future, then it is
+permissible to unreserve a keyword and it is motivated.
+Additionally, if there is a desire for a keyword to be used as an identifier,
+this can in some cases outweigh very hypothetical and speculative language features.*
+
+## Rationale for `pure`
+
+This keyword used to be used for `pure fn`, that is: as an effect.
+
+[applicative]: http://hackage.haskell.org/package/base-4.11.1.0/docs/Control-Applicative.html#t:Applicative
+
+When *generic associated types* (GATs) lands, it is likely that people would
+like to use this in their [applicative functor][applicative] and monad libraries,
+which speaks in favor of unreserving `pure`. This use case explicitly mentioned by [`@ubsan`](https://github.com/ubsan/) who requested that the keyword be unreserved for this purpose.
+
+### Potential drawbacks
+
+Examples / The reasons why we might want to keep `pure` reserved are:
+
+#### 1. Effects
+
+```rust
+pure fn foo(x: Type) -> Type {
+    ...
+}
+```
+
+Here, `pure` denotes a deterministic function -- but we already have `const`
+for more or less the same, and it is unlikely that we would introduce an effect
+(or restriction thereof) that is essentially `const fn` but not entirely.
+So this use case is unlikely to happen.
+
+#### 2. Explicit *`Ok`-wrapping*
+
+```rust
+fn foo() -> Result<i32, Error> {
+    if bar() {
+        pure 0;
+    }
+    ...
+}
+```
+
+desugars into:
+
+```rust
+fn foo() -> Result<i32, Error> {
+    if bar() {
+        return Try::from_ok(0);
+    }
+    ...
+}
+```
+
+[Applicative laws]: https://en.wikibooks.org/wiki/Haskell/Applicative_functors#Applicative_functor_laws
+
+While you might think that Haskell developers would be in favor of this,
+that does not seem to be the case. Haskell developers over at
+`#haskell @ freenode` were not particularly in favor of this use as `pure`
+in this context as `pure` does not respect the [Applicative laws].
+The desugaring is also not particularly obvious when `pure` is used.
+If we did add sugar for explicit `Ok`-wrapping, we'd probably go with something
+other than `pure`.
+
+#### Summary
+
+In both 1. and 2., `pure` can be contextual.
+We also don't think that the drawbacks are significant for `pure`.
+
+## Rationale for `unsize`
```

Comment:
cc @nikomatsakis @mikeyhew

Reply:
`unsized` may be useful for custom DST, so I think it's worth holding onto for now

Reply:
@mikeyhew Can you elaborate a bit?

Reply:
I'd use `dyn` instead of `unsize` for DST.

Reply:
`unsized type` or `unsized struct` might be a desirable syntax for declaring a DST or completely unsized type. I'm just saying it could come in handy as a keyword at some point, and we should hold off on un-reserving it until we have a better idea of how Custom DST will work and it's clear we don't need it. Is there a specific use case for un-reserving it?

Reply:
But `dyn type`/`dyn struct` also works

Reply:
I'd use `dyn` for dynamic existentials, not dynamic sizes, e.g. you could imagine `[T]` expanding to `dyn<n: usize> [T; n]` and `dyn Trait` to `dyn<T: Trait> T`.

Reply:
@eddyb But what about `let vla = [T; dyn size];`?

Reply:
@Centril That's not a size, that's a length. Dynamic existential shorthands (when each bound type/value name is used only once, in an unambiguous way) could be `dyn Trait` and `[T; dyn usize]` (or just `dyn`, since the type can be inferred or specified elsewhere).